[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109647
* #110688

Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

fegin committed:
[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized

Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

wconstab commented on torch/distributed/_composable/replicate.py:
```diff
@@ -11,59 +12,30 @@
-class _ReplicateState:
-    def __init__(self, ignored_modules: Optional[Iterable[torch.nn.Module]]) -> None:
+class _ReplicateState(_State):
+    def __init__(self) -> None:
```
no call to super()? 

what's the reason to subclass _State

wz337 replied:
I believe the reason is that we are using the _State for composable APIs. Correct me if I am wrong.

wz337 commented on torch/distributed/_composable/replicate.py:
```diff
@@ -11,59 +12,30 @@
+    def mark_module(
+        self,
+        module: nn.Module,
+        ignored_modules: Optional[Iterable[torch.nn.Module]],
```
QQ: Should `ignored_params` also be explicitly added to the signature?

fegin replied:
Since the `ignored_modules` is Optional for `replicate`, I think we can leave it as Optional now.

wz337 reviewed (approved):
LGTM

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized"


Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

[ghstack-poisoned]

fegin commented:
@pytorchbot  merge

fegin commented:
@pytorchbot merge
